hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/impl/KeyValueStreamDataChannel.java

```diff
@@ -18,17 +18,100 @@
 
 package org.apache.hadoop.ozone.container.keyvalue.impl;
 
+import com.google.common.base.Preconditions;
 import org.apache.hadoop.hdds.protocol.datanode.proto.ContainerProtos;
+import org.apache.hadoop.hdds.protocol.datanode.proto.ContainerProtos.ContainerCommandRequestProto;
+import org.apache.hadoop.hdds.ratis.ContainerCommandRequestMessage;
+import org.apache.hadoop.hdds.scm.ByteStringConversion;
 import org.apache.hadoop.hdds.scm.container.common.helpers.StorageContainerException;
 import org.apache.hadoop.ozone.container.common.helpers.ContainerMetrics;
 import org.apache.hadoop.ozone.container.common.impl.ContainerData;
+import org.apache.ratis.thirdparty.com.google.protobuf.ByteString;
+import org.apache.ratis.thirdparty.io.netty.buffer.ByteBuf;
+import org.apache.ratis.thirdparty.io.netty.buffer.Unpooled;
 
 import java.io.File;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+
+import org.apache.hadoop.hdds.scm.storage.BlockDataStreamOutput;
 
 /**
  * This class is used to get the DataChannel for streaming.
  */
-class KeyValueStreamDataChannel extends StreamDataChannelBase {
+public class KeyValueStreamDataChannel extends StreamDataChannelBase {
+  private static final ByteBuffer[] EMPTY = {};
+  private static final Function<ByteBuffer, ByteString> CONVERTER
+      // TODO: pass OzoneConfigKeys.OZONE_UNSAFEBYTEOPERATIONS_ENABLED
+      = ByteStringConversion.createByteBufferConversion(false);
+
+  /**
+   * Keep the last {@link Buffers#max} bytes in the buffer
+   * in order to create putBlockRequest
+   * at {@link #closeBuffers(Buffers, WriteMethod)}}.
+   */
+  static class Buffers {
+    private final Deque<ByteBuffer> deque = new LinkedList<>();
+    private final int max;
+    private int length;
+
+    Buffers(int max) {
+      this.max = max;
+    }
+
+    private boolean isExtra(int n) {
+      return length - n >= max;
```

Comment:
Hi @szetszwo , what is the purpose of being greater than 1M?

In the test, writing file size is 128MB, the actual stream's offset in the log was 135272596.
135272596 - (129 \* 1024 \* 1024) = 6292, which is the same as the size of the trespass recorded in logs.

Reply:
> ... what is the purpose of being greater than 1M?

We write putBlockRequestProto at the end of the stream but the putBlockRequestProto should not be written to the file.
The last 1M bytes is kept in the buffer so that we can extract the putBlockRequestProto when the channel is closed.